Make document url mutable and implement location.replace()

stshine opened the pull request:
<!-- Please describe your changes on the following line: -->

---

<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #13413 (github issue number if applicable).

<!-- Either: -->
- [X] There are tests for these changes OR

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/13418)

<!-- Reviewable:end -->

highfive commented:
Heads up! This PR modifies the following files:
 * @kichjang: components/script/dom/webidls/Location.webidl, components/script/dom/location.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify script code, but no tests are modified. Please consider adding a test!

stshine commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 8103751 with merge 8a018c0...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/3353)

stshine commented:
@bors-servo try

KiChjang commented:
@bors-servo retry

bors-servo commented:
:hourglass: Trying commit 8103751 with merge d0e866e...

KiChjang commented:
r=me once you've updated the test expectations. If there are none, you'll need to write one.

@bors-servo delegate+

bors-servo commented:
:v: @stshine can now approve this pull request

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/3354)

jdm commented:
I would be interested in knowing why https://dxr.mozilla.org/servo/source/tests/wpt/web-platform-tests/html/browsers/history/the-location-interface/location_replace.html still fails.

Also, this isn't ready for merging yet - step 1 of https://html.spec.whatwg.org/multipage/browsers.html#dom-location-replace says that we should throw an exception if parsing fails, but that's not included in this PR.

jdm commented:
We should add a test to `location_replace.html` for the parse error exception; we can duplicate the one in `location_assign.html` for that purpose.

stshine commented:
Some  mysterious fails on local machines disappeared.
And Thank for point out the location of the right tests.

stshine commented:
Tried to construct a minimal test case: 
```html
<!DOCTYPE html>
<html>
    <head>
        <style>
         .blank {
             height: 600px;
         }
        </style>
    </head>
    <body>
        <p>Foo</p>
        <div class="blank"></div>
        <a id="x" href="#x">Bar</a>
        <div class="blank"></div>
    </body>
    <script>
     console.log(location.href);
     location.replace("#x");
     setTimeout("eval('console.log(location.href)')", 1000);
    </script>
</html>
```
Try to serve this file with `python3 -m http.server` and visit it.
In firefox the output in console is
> http://localhost:8000/new.html
> http://localhost:8000/new.html#x

while in servo the output is : 
> http://localhost:8000/new.html
> http://localhost:8000/new.html

Tried to dig into the code of `load_url()` and it look like for url that only different in fragment with the url of current document, servo eventually called into `ScriptThread::check_and_scroll_fragment()` in https://github.com/servo/servo/blob/master/components/script/script_thread.rs#L2023 , which only set target state on the target element and scroll to to the frame point.
How am I supposed to fix this?

jdm commented:
I see what you mean. For the purposes of this PR, I propose we add an API to `Document` to set the document's URL and call it from `load_url` before returning when changing the fragment. I've filed #13437 about integrating this with session history, too.

stshine commented:
So... wrap the `url` field in `Cell`? But the `Url` struct does not implement the `Copy` trait.

jdm commented:
DOMRefCell.

stshine commented:
Finally managed to make `document.url` work as intented, but now I got a bunch of erros from wpt runner itself like this:
```
 ▶ ERROR [expected TIMEOUT] /html/browsers/browsing-the-web/scroll-to-fragid/005.html
  └   → Got results from /html/browsers/browsing-the-web/scroll-to-fragid/005.html#test, expected /html/browsers/browsing
-the-web/scroll-to-fragid/005.html                                                                                      
Traceback (most recent call last):
  File "/home/stshine/Programs/servo/tests/wpt/harness/wptrunner/executors/base.py", line 149, in run_test
    result = self.do_test(test)
  File "/home/stshine/Programs/servo/tests/wpt/harness/wptrunner/executors/executorservo.py", line 119, in do_test
    result = self.convert_result(test, self.result_data)
  File "/home/stshine/Programs/servo/tests/wpt/harness/wptrunner/executors/base.py", line 60, in __call__
    (result_url, test.url))
AssertionError: Got results from /html/browsers/browsing-the-web/scroll-to-fragid/005.html#test, expected /html/browsers/
browsing-the-web/scroll-to-fragid/005.html                                                                              
```
Maybe it is not a good idea to set `document.url` ...

stshine commented:
Push my changes up for review. The way I appease borrow checker with RefCell feels quite awkward.

KiChjang commented on components/script/script_thread.rs:
```diff
@@ -1195,7 +1195,7 @@ impl ScriptThread {
         if let Some(fragment) = doc.url().fragment() {
             self.check_and_scroll_fragment(fragment, pipeline, doc);
-        }
+        };
```
nit: Remove this extra `;`.

stshine replied:
This is to appease borrow checker on `RefCell` in `if let`. See https://github.com/rust-lang/rust/issues/22449

KiChjang commented on components/script/script_thread.rs:
```diff
@@ -2017,12 +2018,13 @@ impl ScriptThread {
+                    document.set_url(nurl);
                     return;
-                }
+                } 
```
nit: Remove extra space after `}`.

KiChjang commented on components/script/script_thread.rs:
```diff
@@ -1217,7 +1217,8 @@ impl ScriptThread {
             for it_context in root_context.iter() {
-                let current_url = it_context.active_document().url().to_string();
+                let doc = it_context.active_document();
+                let current_url = doc.url().to_string();
```
This looks like a premature optimization - no code is using the `doc` variable at all, except for `current_url`, so we can just keep the previous code here (i.e. everything on 1 line).

stshine replied:
This is also for appease the borrow checker, we hold the `document` in temp variable so the Ref<Url> from `.url()` will get out of scope before it. And you just can not place the `Ref<T>` in the return value of the function. See https://github.com/rust-lang/rust/issues/23338

KiChjang commented on components/script/script_thread.rs:
```diff
@@ -2017,12 +2018,13 @@ impl ScriptThread {
                 if &url[..Position::AfterQuery] == &nurl[..Position::AfterQuery] &&
                     load_data.method == Method::Get {
                     self.check_and_scroll_fragment(fragment, parent_pipeline_id, document.r());
+                    document.set_url(nurl);
```
We need to leave a note here saying that this is wrong, and should be fixed appropriately via #13437.

stshine replied:
Well, Eventually I think this commit needs to be reverted because a lot of test failed, and I push this commit mostly to notify what actually happened (and to sync between different computers). I hope to get some design suggestions on how the URL fragment should be stored for both the location API and for navigation management.

KiChjang commented on components/script/dom/location.rs:
```diff
@@ -60,6 +60,16 @@ impl LocationMethods for Location {
+        let base_url = self.window.get_url();
+        if let Ok(url) = base_url.join(&url.0) {
+            self.window.load_url(url, true, None);
+        }
```
This code is still not throwing any exceptions.

KiChjang commented on components/script/dom/htmlbodyelement.rs:
```diff
@@ -145,7 +145,9 @@ impl VirtualMethods for HTMLBodyElement {
-                AttrValue::from_url(document_from_node(self).url(), value.into())
+                let doc = document_from_node(self);
+                let val = AttrValue::from_url(&*doc.url(), value.into());
+                val 
```
This creates 2 temporary stack variables for the sake of readability and sacrifices performance. Not a good thing here.

KiChjang commented on components/script/dom/document.rs:
```diff
@@ -363,8 +363,12 @@ impl Document {
+    }
+
+    pub fn set_url(&self, url: &Url) {
+        *self.url.borrow_mut() = url.clone();
```
Why is a `DOMRefCell` necessary? Does returning a mutable reference to `self.url` not work?

Also, we need to leave a FIXME here pointing to #13437.

stshine replied:
That would need to obtain the `document` as mutable, which I did not figure out how to do that.

KiChjang replied:
```rust
pub fn url_mut(&mut self) {
    &mut self.url
}
```

Ms2ger replied:
No, no, NO. There are no `&mut` references to DOM objects.

KiChjang replied:
This could really take in a `Url` instead of a `&Url`.

KiChjang commented on components/script/dom/canvasrenderingcontext2d.rs:
```diff
@@ -240,7 +240,9 @@ impl CanvasRenderingContext2D {
-                        url.origin() == node.owner_doc().url().origin()
+                        let doc = node.owner_doc();
+                        let result = url.origin() == doc.url().origin();
+                        result
```
This is adding 2 temporary stack variables here for no gain in readability and added verbosity. Revert.

KiChjang replied:
Needs comment linking to rust-lang/rust#37785.

stshine commented:
Maybe make `document.url` mutable is feasible, just that there are still some bugs in the implementation. I will reply when I figure more things out.

jdm commented:
@stshine Yeah, I'm beginning to reconsider my suggestion of making the URL mutable. Here's another idea - what if we add a `fragment: DOMRefCell<String>` member to Document and mutate that when appropriate. Then we need to make Location's `get_url` method return the url with the fragment appended to it. How does that sound?

stshine commented:
I'm not oppose to that. Since the hash of a URL indicate location inside a webpage, giving it some special treatment is really not a bad idea that it may enable some future optimization. That said, I'm not familiar with javascript apis that obtain current URL, but if there are only `location` and `document.URL`, it is not hard to implement in both options, but it may need a additional clone of `Url` if stored separately. And consider the situation that the web is getting more app link investigation on  the behavior of some modern javascript libraries like react-router may help make the decision.

stshine commented:
Looks like the problem of the current implementation is a timing issue. Giving the following test:
```html
<!DOCTYPE html>
<html>
    <head>
    </head>
    <body>
        <p>Foo</p>
    </body>
    <script>
     console.log(location.href);
     location.replace("#x");
     eval("console.log(location.href)");
     setTimeout("console.log(location.href)", 1);
    </script>
</html>
```
The ouput is:
> http://localhost:8000/new.html
> http://localhost:8000/new.html
> http://localhost:8000/new.html#x

Really don't know how to deal with this, please give some suggestions...

jdm commented:
When you say the current implementation, I assume you mean the mutable document URL? That makes sense; our implementation of navigation is entirely asynchronous, but the specification says that step 5 of https://html.spec.whatwg.org/multipage/browsers.html#navigate (ie. is the URL the same except for the fragment?) should happen synchronously. We should fix that in a separate issue, I think.

stshine commented:
Since some tests like `location_replace.html` do not restore the url fragment and `hashchange` event has not been implemented we need to restore the url in test report.
Blocks on https://github.com/w3c/wptrunner/pull/214

jdm commented on components/script/dom/location.rs:
```diff
@@ -78,7 +84,12 @@ impl LocationMethods for Location {
     // https://html.spec.whatwg.org/multipage/#dom-location-hash
     fn SetHash(&self, mut value: USVString) {
         if value.0.is_empty() {
-            value = USVString("#".to_owned());
```
I think this change will make us fail tests? The spec has text specifically about how Location.hash's setter is supposed to behave when passed an empty string, and it's supposed to have # appended to it afterwards.

jdm commented on components/script/dom/location.rs:
```diff
@@ -78,7 +84,12 @@ impl LocationMethods for Location {
+            let mut url = self.get_url();
+            url.set_fragment(None);
+            doc.set_url(&url);
+            self.window.check_and_scroll_fragment("");
```
Why is this code necessary given the changes in Window::load_url?

jdm commented on components/script/script_thread.rs:
```diff
@@ -2143,8 +2083,6 @@ impl ScriptThread {
         // https://html.spec.whatwg.org/multipage/#the-end steps 3-4.
         document.process_deferred_scripts();
-
-        window.set_fragment_name(final_url.fragment().map(str::to_owned));
```
Hmm, I wonder if the goal of this code was to scroll to the fragment once the load event was dispatched (although I bet that relied on a resize event being processed afterwards, and we probably broke that since this code was written :P)

stshine replied:
Yes, as my experiment on firefox and chrome, both browsers do not scroll to the fragment in the url after resize; instead, they kind of remeber the previous scroll point and keep to that point after resize, so I guess it is appropriate to remove this mechanism, since this behavior is kind of weird; that said, I have not be able to find the specification of the behavior of resize.

jdm commented on components/script/dom/window.rs:
```diff
@@ -1414,6 +1448,16 @@ impl Window {
         let referrer_policy = referrer_policy.or(doc.get_referrer_policy());
 
+
+        if let Some(fragment) = url.fragment() {
```
Let's add a link to https://html.spec.whatwg.org/multipage/#navigating-across-documents and a `// Step 7` comment.

stshine replied:
Will do it.

stshine replied:
Oh, actually it is step 5.

jdm commented on components/script/dom/window.rs:
```diff
@@ -1414,6 +1448,16 @@ impl Window {
 
+
+        if let Some(fragment) = url.fragment() {
+            let old_url = (*doc.url()).clone();
```
I don't believe the clone is necessary here.

stshine replied:
I'll study this after rebase.

jdm commented on components/script/dom/window.rs:
```diff
@@ -1056,6 +1053,43 @@ impl Window {
     }
 
+    pub fn check_and_scroll_fragment(&self, fragment: &str) {
+        // TODO(stshine): if fragment is empty, scroll to the top.
```
We could use `Window::GetBody()` as the scroll target in this case.

jdm commented on components/script/dom/location.rs:
```diff
@@ -57,7 +60,10 @@ impl LocationMethods for Location {
+        self.window.main_thread_script_chan().send(
+            MainThreadScriptMsg::Navigate(self.window.pipeline_id(),
+                                          LoadData::new(self.get_url(), None, Some(self.get_url())),
+                                          true)).unwrap();
```
I would prefer to add an argument to `load_url` that indicates that we should ignore the fragment check.

stshine replied:
My previous plan is: for navigate request that should always happen, just send the message to `MainThreadScriptMsg`; otherwise call `window.load_url()`. But I bet you have a reason for this since I have no idea about the character of `ScriptThread`.

KiChjang commented:
w3c/wptrunner#214 has been merged.

KiChjang commented:
@stshine Are you still working on this?

stshine commented:
@KiChjang Oops, Sorry! Work is still quite busy, and currently I spend most of my free time trying to solve my floating issue #13683. Again it is harder than I think, but I believe I can work out a PR in this week, and after which I will continue to work on this. Since this PR can not pass until wptrunner is merged from upstream next time, I did not take this as of high priority.
 I did not expect it would take this long, and I confess that I did not respond to the reviews in fear of cannot giving out constructive comments right now, Sorry @jdm.

stshine commented:
First, @jdm sorry for letting you review a wip work. Some changes are here to try to  keep the fragment invariant after `location.hash = ""`, which proves to be not enough for our test. I will rebase on to master and clean this up.
Another topic for this PR is there are some concerns about making `document.url` refcounted, see #9782 and #13790 . Would you like to give some comments on relationship between that and this PR, and if we are going to use refcounting eventually, how do we keep the safety and synchronization since fragment is possible to change frequently?

KiChjang commented:
@stshine Don't worry about refcounting Urls for the moment, this is completely out of scope of Location.replace.

jdm commented:
That being said, it's an interesting point to consider for the discussion in #13790.

stshine commented:
This pull request is ready for review. 
Still I kind of hope to integrate `scroll_fragment_point()` into other scroll methods.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #14013) made this pull request unmergeable. Please resolve the merge conflicts.

KiChjang commented on components/script/dom/document.rs:
```diff
@@ -376,8 +376,12 @@ impl Document {
-    pub fn url(&self) -> &Url {
-        &self.url
+    pub fn url(&self) -> Ref<Url> {
+        self.url.borrow()
```
This can just use a `.clone()` for now, returning `Ref<Url>` isn't the scope of this PR.

stshine replied:
Oh but @jdm seems not like clones of `Url`... @jdm What is your opinion here?

stshine replied:
Well since the previously `document.url()` returns a reference, now change it to return the actual struct would results changes to a lot of places(roughly 50, and they need to be changed again if we implement refcounting later). If you can assure that is OK I will make the change.

KiChjang replied:
Oh, didn't realize the previous code was returning &Url, nevermind.

KiChjang reviewed (commented):
Partial review, this is removing quite a bit of code, so I need to pull up the spec and ensure everything is in order.

KiChjang commented on components/script/dom/window.rs:
```diff
@@ -1356,7 +1402,9 @@ impl Window {
-        (*self.Document().url()).clone()
+        let doc = self.Document();
+        let val = (*doc.url()).clone();
+        val
```
Can this not be reverted?

stshine replied:
I am sure to say that there is no workaround unless we add a method to `Document` that returns a clone of the `Document.url`.

KiChjang replied:
Actually, this is very odd, since document.url() returns a `&Url`, but in this function we purposefully deref it just to return a `Url`. I suppose this is the status quo until #13790 is solved.

Regardless of return values, we still need to figure out _why_ exactly the lifetimes aren't working out as expected. It is not normal to see how using a `&Url` causes no borrrowck issues but a `Ref<Url>` with an equivalent lifetime causes a borrow error.

KiChjang replied:
I've opened up rust-lang/rust#37785 for this. In the meantime, add a comment linking to the relevant rust-lang issue and mention that this is a temporary fix to appease borrowck.

nox replied:
At the very least, the `val` variable is unneeded, AFAIK.

KiChjang replied:
@nox I pulled his branch and tried that too, but to no avail. I hear that it's because of dropck.

KiChjang replied:
Needs comment linking to rust-lang/rust#37785.

KiChjang commented on components/script/dom/xmlhttprequest.rs:
```diff
@@ -159,7 +159,8 @@ impl XMLHttpRequest {
         let (referrer_url, referrer_policy) = if let Some(window) = global.downcast::<Window>() {
             let document = window.Document();
-            (Some(document.url().clone()), document.get_referrer_policy())
+            let val = (Some((*document.url()).clone()), document.get_referrer_policy());
```
nit: Extra parentheses around Some. Also, can this not be reverted?

KiChjang commented on components/script/dom/window.rs:
```diff
@@ -989,6 +987,51 @@ impl Window {
     }
 
+    /// https://html.spec.whatwg.org/multipage/#scroll-to-the-fragment-identifier
+    pub fn check_and_scroll_fragment(&self, fragment: &str) {
```
Why is this code moved here instead of document.rs?

stshine replied:
Will move.

KiChjang commented on components/script/dom/window.rs:
```diff
@@ -989,6 +987,51 @@ impl Window {
+
+        let target = target.r().map(|e| &*e).or_else(
+            // Step 2
+            || if fragment == "" {
```
nit: Closure bars should be in the previous line. Also, `if fragment.is_empty()`.

KiChjang commented on components/script/dom/window.rs:
```diff
@@ -989,6 +987,51 @@ impl Window {
+        let target = target.r().map(|e| &*e).or_else(
+            // Step 2
+            || if fragment == "" {
+                body.r().map(|b| b.upcast::<Element>())
```
This looks wrong. When `fragment` is the empty string, the `find_fragment_node` algorithm points to the document node instead of the document's body.

stshine replied:
This for scrolling to the top when the fragment is empty, and the document node does not seems to have a bounding box.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #14023) made this pull request unmergeable. Please resolve the merge conflicts.

stshine commented on components/script/dom/window.rs:
```diff
@@ -989,6 +987,51 @@ impl Window {
+        let target = doc.find_fragment_node(fragment);
+
+        // Step 1
+        doc.set_target_element(target.r().map(|e| &*e));
```
It is here because it is a Option<Root<T>>. You can see a lot of usage of `.r()` for this.

KiChjang commented on components/script/dom/document.rs:
```diff
@@ -574,6 +578,28 @@ impl Document {
+        self.set_target_element(target.r());
+
+        let target = target.r().or_else(||
+                                        // Step 2
```
nit: Weird indentation.

KiChjang commented on components/script/dom/xmlhttprequest.rs:
```diff
@@ -158,7 +158,8 @@ impl XMLHttpRequest {
         let (referrer_url, referrer_policy) = if let Some(window) = global.downcast::<Window>() {
             let document = window.Document();
-            (Some(document.url().clone()), document.get_referrer_policy())
+            let val = (Some(document.url().clone()), document.get_referrer_policy());
```
Needs comment linking to rust-lang/rust#37785.

KiChjang commented on components/script/script_thread.rs:
```diff
@@ -1252,7 +1238,8 @@ impl ScriptThread {
         for (_, document) in self.documents.borrow().iter() {
-            let current_url = document.url().as_str();
+            let url = document.url();
+            let current_url = url.as_str();
```
I'm assuming this is also because of rust-lang/rust#37785?

KiChjang commented on components/script/dom/htmlbodyelement.rs:
```diff
@@ -146,7 +146,9 @@ impl VirtualMethods for HTMLBodyElement {
             local_name!("text") => AttrValue::from_legacy_color(value.into()),
             local_name!("background") => {
-                AttrValue::from_url(document_from_node(self).url(), value.into())
+                let doc = document_from_node(self);
```
Needs comment linking to rust-lang/rust#37785.

KiChjang commented on components/script/dom/window.rs:
```diff
@@ -1322,10 +1341,21 @@ impl Window {
+            // Step 5
+            if let Some(fragment) = url.fragment() {
+                doc.check_and_scroll_fragment(fragment);
+                doc.set_url(url.clone());
```
Does borrowck complain if you remove the .clone()?

stshine replied:
Yes, it is borrowed by `if let`.

KiChjang commented on components/script/dom/document.rs:
```diff
@@ -574,6 +578,28 @@ impl Document {
+    /// https://html.spec.whatwg.org/multipage/#scroll-to-the-fragment-identifier
+    pub fn check_and_scroll_fragment(&self, fragment: &str) {
+        let body = self.GetBody();
+        let target = self.find_fragment_node(fragment);
```
Does .r() not work here?

stshine replied:
No... you have to keep the Root<T> on the stack.

KiChjang commented on components/script/dom/document.rs:
```diff
@@ -574,6 +578,28 @@ impl Document {
+        let target = target.r().or_else(||
+                                        // Step 2
+                                        if fragment.is_empty() {
+                                            body.r().map(|b| b.upcast::<Element>())
```
So you mean `scroll_fragment_point` would spit out an error if `target` is a `Document` element? This means that you'll need to check whether `fragment == "top"` as well, since `find_fragment_node` will return a `Document`.

stshine replied:
So the origin of the stacking context space where all bounding box live in is: `                LogicalPoint::zero(writing_mode).to_physical(writing_mode, viewport_size)`, which is not (0, 0) when the writing mode is not `horizontal-lr`. But currently servo does not care about writing mode too much, So I will simply use (0, 0) to scroll to the top.

stshine commented:
Comments addressed, and integrated `scroll_fragment_point()` into `perform_a_scroll()`.

KiChjang commented on components/script/dom/document.rs:
```diff
@@ -578,6 +578,46 @@ impl Document {
+        self.set_target_element(target.r());
+
+        let point;
+        if fragment.is_empty() || fragment.to_lowercase() == "top" {
```
`let point = if fragment.is_empty //...`

stshine commented:
Comment addressed.

KiChjang commented on components/script/dom/canvasrenderingcontext2d.rs:
```diff
@@ -239,7 +239,10 @@ impl CanvasRenderingContext2D {
                         // the entry settings object, but for now check it against the canvas' doc.
                         let node: &Node = &*self.canvas.upcast();
-                        url.origin() == node.owner_doc().url().origin()
+                        //FIXME: https://github.com/rust-lang/rust/issues/37785
```
nit: Space between / and F, and for the other comments as well.

stshine commented:
Updated according to comment.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #14246) made this pull request unmergeable. Please resolve the merge conflicts.

stshine committed:
Make 'Document.url' field mutable

Use DOMRefCell as the type of the url field.

stshine committed:
Move fragment navigation into Document object

Move the `check_and_scroll_fragment()` method into Document, make the
mothod set the fragment of url after navigation, and use the
`perform_a_scroll()` method to scroll rather than an individual
method. Also removes the broken `Window.fragment` fields.

stshine committed:
Implement Location.replace

stshine committed:
Remove redundant url clones

They are now redundant since now document.url() returns a struct rather
than a reference.

stshine committed:
Minor fixes and update test expectations

stshine commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit f58ccff with merge d20243e...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/921)

stshine commented:
Rebased and updated test expectations.

KiChjang commented on components/script/dom/document.rs:
```diff
@@ -192,7 +192,7 @@ pub struct Document {
     encoding: Cell<EncodingRef>,
     is_html_document: bool,
-    url: ServoUrl,
+    url: DOMRefCell<ServoUrl>,
```
Using the new ServoUrl struct, having a DOMRefCell is not necessary anymore.

KiChjang commented:
r=me after nits.

stshine commented on components/script/script_thread.rs:
```diff
@@ -1263,10 +1249,10 @@ impl ScriptThread {
 
             reports.push(Report {
-                path: path![format!("url({})", current_url), "dom-tree"],
+                path: path![format!("url({})", current_url.as_str()), "dom-tree"],
```
Oh this is because now `document.url()` return a value while `as_str()` returns a reference, so we have to keep the value on the stack.

stshine commented:
@bors-servo r=KiChjang

bors-servo commented:
:pushpin: Commit 3028cbe has been approved by `KiChjang`

<!-- @bors-servo r=KiChjang 3028cbe06b6560619b9f7961d32e0f8e4e4ecea4 -->

bors-servo commented:
:bulb: This pull request was already approved, no need to approve it again.

- There's another pull request that is currently being tested, blocking this pull request: #14256

bors-servo commented:
:pushpin: Commit 3028cbe has been approved by `KiChjang`

<!-- @bors-servo r=KiChjang 3028cbe06b6560619b9f7961d32e0f8e4e4ecea4 -->

bors-servo commented:
:hourglass: Testing commit 3028cbe with merge b814390...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/970)

KiChjang commented:
@bors-servo retry
* #13133

bors-servo commented:
:hourglass: Testing commit 3028cbe with merge 1d1b742...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/978)

stshine commented:
Hmm I have to take a look at why /old-tests/submission/Opera/script_scheduling/028.html passed in the first try. Hope it won't become a new intermittent...

stshine commented:
@bors-servo r=KiChjang
#11100
There are two sub tests in old-tests/submission/Opera/script_scheduling/028.html while only one get passed, silly me.

bors-servo commented:
:pushpin: Commit dc4bc99 has been approved by `KiChjang`

<!-- @bors-servo r=KiChjang dc4bc99fcc48f2b58c34cfea3c33712455b4d931 -->

emilio commented on components/script/dom/document.rs:
```diff
@@ -596,6 +600,45 @@ impl Document {
+        // Step 1
+        self.set_target_element(target.r());
+
+        let point = if fragment.is_empty() || fragment.to_lowercase() == "top" {
```
nit: use `eq_ignore_ascii_case("top")` instead.

stshine replied:
Darnnnnn, suddenly found that I should just use `Cell` for `ServoUrl` rather than `DOMRefCell`. Sadface :-(

stshine replied:
ignore my previous comment.

emilio reviewed (commented):
If it doesn't merge at the first try please do that, otherwise probably worth opening an followup with `E-easy` or similar :).

bors-servo commented:
:hourglass: Testing commit dc4bc99 with merge b8cb92a...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/965)

highfive commented:
▶ FAIL [expected PASS] /_mozilla/css/incremental_trailing_whitespace_a.html
      └   → /_mozilla/css/incremental_t</span><span class="stdout">railing_whitespace_a.html da54041086f2975f8e3c776d2283ad6609e6862a
    /_mozilla/css/incremental_trailing_whitespace_ref.html f5d0147c8bddbdc772a7e5e86c7c9e433fcd486b
    Testing da54041086f2975f8e3c776d2283ad6609e6862a == f5d0147c8bddbdc772a7e5e86c7c9e433fcd486b

stshine commented:
@bors-servo r=KiChjang
#10473 
comment addressed.

bors-servo commented:
:pushpin: Commit 5287e70 has been approved by `KiChjang`

<!-- @bors-servo r=KiChjang 5287e708291770617fe75d4f8fd59593b84b80b1 -->

bors-servo commented:
:hourglass: Testing commit 5287e70 with merge debad8a...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/977)

KiChjang commented:
@bors-servo retry
* #12797

bors-servo commented:
:hourglass: Testing commit 5287e70 with merge 72e4c6d...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/4228), [arm64](http://build.servo.org/builders/arm64/builds/4236), [linux-dev](http://build.servo.org/builders/linux-dev/builds/4322), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/984), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/984), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/4305), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/4310), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/1003), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/1008), [windows-dev](http://build.servo.org/builders/windows-dev/builds/2256)

stshine commented:
Waaaaoh! Finally this PR got merged before it has 150 comments. Thanks @KiChjang and @jdm for review, please forgive my single-threaded head :)

emilio commented:
Thank _you_ for being so persistent :)

On Sun, Nov 20, 2016 at 04:04:52AM -0800, stshine wrote:
> Waaaaoh! Finally this PR got merged before it has 150 comments. Thanks @KiChjang and @jdm for review, please forgive my single-threaded head :)
> 
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/servo/servo/pull/13418#issuecomment-261774209
